File: LeetCode/238_Product of Array_Except_Self.py
# def productExceptSelf(nums):
#     result = [1]*(len(nums))
#     prefix = 1
#     for i in range(len(nums)):
#         result[i] = prefix
#         prefix *= nums[i]
#         print(f'index: {i}, result[{i}]: {result[i]} prefix: {prefix}')
#         print(result)
#         print('')
#     postfix = 1
#     for i in range(len(nums)-1, -1 , -1):
#         result[i] *= postfix
#         postfix *= nums[i]
#         print(f'index: {i}, result[{i}]: {result[i]} postfix: {postfix}')
#         print(result)
#         print('')
        
def productExceptSelf(nums):
    result = [1]*len(nums)
    prefix = 1
    for i in range(len(nums)):
        result[i] = prefix
        prefix *= nums[i]
    
    postfix = 1
    for i in range(len(nums)-1, -1 ,-1):
        result[i] *= postfix
        postfix *= nums[i]
    print(result)
        

        



# a = [1,2,3,4]
# a = [1,2,3,4,5,6]
a = [-1, 1, 0, -3, 3]
productExceptSelf(a)
# a.remove(a[1])
# print(a)
# print(a[2:] + a[0::-1])
# print(a[1:len(a)])
# print(a[2:]+ a[0::-1])
# print(a[1:]+ a[-1::-1])

File: LeetCode/238_Product_of_Array_Except_Self.js
/**
 * Input: nums = [1,2,3,4]
 * Output: [24,12,8,6]
 * 
 * Input: nums = [-1,1,0,-3,3]
 * Output: [0,0,9,0,0]
 * 
 */

var productExceptSelf = function(nums) {
    var arr = Array(nums.length).fill(1)
    console.log(arr)
    var prefix = 1;

    for(let i = 0; i < nums.length; i++) {
        arr[i] = prefix;
        prefix *= nums[i];
    };
    var post = 1
    for (let i = nums.length-1; i >= 0; i--) {
        arr[i] *= post;
        post *= nums[i]
    };
    console.log(arr)
};


const nums = [1,2,3,4]
productExceptSelf(nums)